[GStreamer][MediaStream] Additional fixes for disabled video track handling

philn opened the pull request:
#### 853b286ebdf4b2951746c4f93ab22e051fe4cb47
<pre>
[GStreamer][MediaStream] Additional fixes for disabled video track handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=243776">https://bugs.webkit.org/show_bug.cgi?id=243776</a>

Reviewed by Xabier Rodriguez-Calvar.

The internal audio/video appsrc elements are now named with monotonically increasing id in order to
ease debugging a bit.

As the element can receive the trackEnabledChanged notification from a non-main thread, we need to
keep track of the enabled state in a boolean flag, so that we can start pushing black frames as soon
as the notification was received.

The black frame default size is now fixed to 320x240 unless a size was previously configured, eg.
when the track was enabled. The previous sample caps can&apos;t be used because they might describe an
RTP payload and not a raw video format.

* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:

Canonical link: <a href="https://commits.webkit.org/253467@main">https://commits.webkit.org/253467@main</a>
</pre>

ceyusa commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -241,7 +246,7 @@ class InternalSource final : public MediaStreamTrackPrivate::Observer,
 
         if (m_track.isVideo() && drop)
-            drop = doCapsHaveType(caps, "video/x-raw") || GST_BUFFER_FLAG_IS_SET(buffer, GST_BUFFER_FLAG_DELTA_UNIT);
+            drop = doCapsHaveType(caps, "video") || GST_BUFFER_FLAG_IS_SET(buffer, GST_BUFFER_FLAG_DELTA_UNIT);
```
if track isVideo, perhaps checking its caps -so generically- might be spurious.

philn replied:
I'm not sure what you mean here, the caps can be either `video/x-raw` or `application/x-rtp media=video`.

ceyusa replied:
oh, ok! I didn't figured out the rtp side. Sorry for the noise.

ceyusa commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -370,7 +383,7 @@ class InternalSource final : public MediaStreamTrackPrivate::Observer,
             memset(data.data() + yOffset, 128, data.size() - yOffset);
         }
-        gst_buffer_add_video_meta_full(buffer.get(), GST_VIDEO_FRAME_FLAG_NONE, GST_VIDEO_FORMAT_I420, m_lastKnownSize.width(), m_lastKnownSize.height(), 3, info.offset, info.stride);
+        gst_buffer_add_video_meta_full(buffer.get(), GST_VIDEO_FRAME_FLAG_NONE, GST_VIDEO_FORMAT_I420, width, height, 3, info.offset, info.stride);
```
if this a software filled buffer, perhaps there's no need of video meta.

philn replied:
I dunno, we have offsets/stride info, I see no reason not to include it in a meta. Or would that have a perf impact?

ceyusa replied:
perf impact, minimal at most, I  just  wondered about removing non-functional code. But it doesn't harm either.

philn replied:
Ok, I think we can revisit this part later on. It's a bit of a corner-case anyway because this code is about disabled video track handling. Most of the time users want an enabled track :)

calvaris reviewed (approved):
After clarifying Víctor's concerns, you can land.

philn committed:
[GStreamer][MediaStream] Additional fixes for disabled video track handling
https://bugs.webkit.org/show_bug.cgi?id=243776

Reviewed by Xabier Rodriguez-Calvar.

The internal audio/video appsrc elements are now named with monotonically increasing id in order to
ease debugging a bit.

As the element can receive the trackEnabledChanged notification from a non-main thread, we need to
keep track of the enabled state in a boolean flag, so that we can start pushing black frames as soon
as the notification was received.

The black frame default size is now fixed to 320x240 unless a size was previously configured, eg.
when the track was enabled. The previous sample caps can't be used because they might describe an
RTP payload and not a raw video format.

* Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:

Canonical link: https://commits.webkit.org/253467@main

blino commented on Source/WebCore/platform/mediastream/gstreamer/GStreamerMediaStreamSource.cpp:
```diff
@@ -380,6 +393,7 @@ class InternalSource final : public MediaStreamTrackPrivate::Observer,
         VideoFrameTimeMetadata metadata;
         metadata.captureTime = MonotonicTime::now().secondsSinceEpoch();
         auto* buffer = webkitGstBufferSetVideoFrameTimeMetadata(gst_sample_get_buffer(m_blackFrame.get()), metadata);
+        GST_BUFFER_DTS(buffer) = GST_BUFFER_PTS(buffer) = gst_element_get_current_running_time(m_parent);
```
This requires GStreamer 1.18 for `gst_element_get_current_running_time()`, while `GStreamerChecks.cmake` only requires GStreamer 1.14.0 for now.

Note also the 1.16 TODO comment below.

philn replied:
If it's easy to vendor in GStreamerCommon let's do it, otherwise ifdef. Do you want me to follow-up on that or you prefer to?

blino replied:
I'd prefer to let you do it, I am not sure what is the best behavior on GStreamer < 1.18

philn replied:
PTS/DTS handling is a rather stable concept nowadays, so I think it's safe to vendor that get_current_time API until we can bump the dep to 1.18. I'll send a MR. Thanks for raising this issue!

philn replied:
https://github.com/WebKit/WebKit/pull/3909

blino replied:
No problem.
For the record, this was to fix build on Ubuntu 20.04.
